### PR TITLE
CachePlugin typings

### PR DESCRIPTION
**Pull request by Legends**
**What kind of change does this PR introduce?**
adds typings to CachePlugin

**Did you add tests for your changes?**
N/A

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
N/A

### CODE REVIEW

File: lib/CachePlugin.js [hunk 1]
```diff
@@ -6,18 +6,39 @@
...
 	apply(compiler) {
 		if (Array.isArray(compiler.compilers)) {
 			compiler.compilers.forEach((c, idx) => {
 				new CachePlugin((this.cache[idx] = this.cache[idx] || {})).apply(c);
 			});
 		} else {
+			/**
+			 *
+			 * @param {Compiler} compiler webpack compiler
+			 * @param {Object} cache cache object, where the modules are cached
```
Comment (mohsen1) [inline lib/CachePlugin.js]: If you can type the cache object it would be more type safe

**Comment by Legends** [comment]
@mohsen1:  Hmmm ...
What can we do about this?

> Property 'compilers' does not exist on type 'Compiler'

File: lib/CachePlugin.js (hunk 1 above)
Comment (Legends) [reply lib/CachePlugin.js]: Where is it defined..the `cache` object?

File: lib/CachePlugin.js [hunk 2]
```diff
@@ -6,18 +6,39 @@
...
+/** @typedef {import("./Compiler.js")} Compiler */
+
+/**
+ * Adds a cache to the compiler, where modules are cached.
+You can pass a cache object, where the modules are cached. Otherwise one is created per plugin instance.
+ */
 class CachePlugin {
+	/**
+	 *
+	 * @param {Object} [cache] cache object, where the modules are cached
```
[the remark this answers is not in the document]
Comment (Legends) [reply lib/CachePlugin.js]: @ooflorent  I thought square brackets means optional, like mentioned [here - jsDoc-Syntax](http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values)

File: lib/CachePlugin.js (hunk 1 above)
Comment (Legends) [reply lib/CachePlugin.js]: Ok, I guess `Object` should be fine then...

File: lib/CachePlugin.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ooflorent) [reply lib/CachePlugin.js]: We previously agreed on using `=` instead of brackets

[the remark this answers is not in the document]
Comment (ooflorent) [reply lib/CachePlugin.js]: More info here: https://github.com/webpack/webpack/issues/7122

File: lib/CachePlugin.js (hunk 1 above)
Comment (Legends) [reply lib/CachePlugin.js]: If the `cache` can cache other things beside `modules` [this part](https://github.com/webpack/docs/wiki/internal-webpack-plugins#cacheplugincache) could be adapted..

File: lib/CachePlugin.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Legends) [reply lib/CachePlugin.js]: Good to know, thanks...will update the PR

**Comment by Legends** [comment]
@ooflorent What about 

> Property 'compilers' does not exist on type 'Compiler'

**Comment by Legends** [comment]
@ooflorent, @mohsen1 

> You could change your param to {Compiler | MultiCompiler}..

**What I don't understand:** 
Is the `compiler` param passed to the `apply` method of type `MultiCompiler` or `Compiler`? Both inherit from `Tapable`.

Because of the `compilers` property I would say `MultiCompiler`.

**Comment by Legends** [comment]
I've set it to `@param {TODO} compiler Webpack Compiler` like @mohsen1  did in `DefinPlugin`.

File: lib/CachePlugin.js [hunk 3]
```diff
@@ -6,18 +6,40 @@
 
 const asyncLib = require("neo-async");
 
+/** @typedef {import("./Compiler.js")} Compiler */
+/** @typedef {import("./MultiCompiler.js")} MultiCompiler */
```
Comment (ooflorent) [inline lib/CachePlugin.js]: These 2 imports are no longer needed

File: lib/CachePlugin.js [hunk 4]
```diff
@@ -6,18 +6,40 @@
...
+
+/**
+ * Adds a cache to the compiler, where modules are cached.
+You can pass a cache object, where the modules are cached. Otherwise one is created per plugin instance.
```
Comment (ooflorent) [inline lib/CachePlugin.js]: nit: Please add `*`

File: lib/CachePlugin.js (hunk 3 above)
Comment (Legends) [reply lib/CachePlugin.js]: A yes, totally forgot the remove them. thanks..

File: lib/CachePlugin.js [hunk 5]
```diff
@@ -6,18 +6,37 @@
...
+	 * @param {Object=} cache cache object
+	 */
 	constructor(cache) {
 		this.cache = cache || {};
 		this.FS_ACCURACY = 2000;
 	}
 
+	/**
+	 *
+	 * @param {TODO} compiler Webpack Compiler
```
Comment (sokra) [inline lib/CachePlugin.js]: TODO -> Compiler

File: lib/CachePlugin.js [hunk 6]
```diff
@@ -6,18 +6,37 @@
...
+	 */
 	apply(compiler) {
 		if (Array.isArray(compiler.compilers)) {
 			compiler.compilers.forEach((c, idx) => {
 				new CachePlugin((this.cache[idx] = this.cache[idx] || {})).apply(c);
 			});
 		} else {
+			/**
+			 *
+			 * @param {TODO} compiler webpack compiler
```
Comment (sokra) [inline lib/CachePlugin.js]: TODO -> Compiler

File: lib/CachePlugin.js (hunk 5 above)
Comment (Legends) [reply lib/CachePlugin.js]: @sokra 
TS complains: Property 'compilers' does not exist on type 'Compiler'.
Still change the type to `Compilers`?

**Comment by Legends** [comment]
Because of deleted local branch adding new [PR](https://github.com/webpack/webpack/pull/7570)

**Pull request closed without merging**